lib/Model/Importer/DebugInfo/PDBImporter.cpp

```diff
@@ -0,0 +1,1093 @@
+  // Parse debug info and populate types to Model.
+  const codeview::DebugInfo *DebugInfo;
+  StringRef PDBFilePath;
+  std::unique_ptr<IPDBSession> Session;
+  NativeSession *TheNativeSession;
+
+  auto EC = TheBinary.getDebugPDBInfo(DebugInfo, PDBFilePath);
+  if (not EC and DebugInfo != nullptr and not PDBFilePath.empty()) {
+    if (llvm::sys::fs::exists(PDBFilePath)) {
+      auto Err = loadDataForPDB(PDB_ReaderType::Native, PDBFilePath, Session);
+      if (not Err) {
+        TheNativeSession = static_cast<NativeSession *>(Session.get());
+        // TODO: We are using the static_cast due to lack of an LLVM RTTI
+        // support for this. Once it is improved in LLVM, we should avoid this.
+        if (Session->getLoadAddress() != TheNativeSession->getLoadAddress())
+          revng_abort();
+
+        ThePDBFile = &TheNativeSession->getPDBFile();
+      } else {
+        revng_log(DILogger, "Unable to read PDB file.");
+      }
+    } else {
+      // Usualy the PDB files will be generated on a different machine,
+      // so the location read from the debug directory wont be up to date. At
+      // first try to find it in the current `.` dir.
+      // TODO: Introduce a `cl::opt` that will be used for providing a
+      // whitelisted  paths for finding PDB files.
+      auto PDBFileNameOnly = PDBFilePath.substr(PDBFilePath.find_last_of('\\')
+                                                + 1);
+      if (llvm::sys::fs::exists(PDBFileNameOnly.str())) {
+        auto Err = loadDataForPDB(PDB_ReaderType::Native,
+                                  PDBFileNameOnly.str(),
+                                  Session);
+        if (not Err) {
+          TheNativeSession = static_cast<NativeSession *>(Session.get());
+          // TODO: The same as the TODO marker above.
+          if (Session->getLoadAddress() != TheNativeSession->getLoadAddress())
+            revng_abort();
+
+          ThePDBFile = &TheNativeSession->getPDBFile();
+        } else {
+          revng_log(DILogger, "Unable to read PDB file.");
+        }
+      } else {
+        revng_log(DILogger, "Unable to find PDB file.");
+      }
+
+      if (ThePDBFile) {
+        // According to llvm/docs/PDB/PdbStream.rst, the `Signature`was never
+        // used the way as it was the initial idea. Insead, GUID is a 128-bit
+        // identifier guaranteed to be unique ID for both executable and
+        // corresponding PDB.
+        codeview::GUID GUIDFromExe;
+        std::copy(std::begin(DebugInfo->PDB70.Signature),
+                  std::end(DebugInfo->PDB70.Signature),
+                  std::begin(GUIDFromExe.Guid));
+        auto PDBInfoStrm = ThePDBFile->getPDBInfoStream();
+        if (!PDBInfoStrm)
+          revng_log(DILogger, "No PDB Info stream found.");
+        else {
+          codeview::GUID GUIDFromPDBFIle = PDBInfoStrm->getGuid();
+          if (GUIDFromExe != GUIDFromPDBFIle)
+            revng_log(DILogger, "Signatures from exe and PDB file mismatch.");
```

Comment:
I guess this is not just for `exe` but DLL too?

Reply:
Right.